### ISSUE TITLE
HIdden tags 

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#1366]
Introductory note: IMO there should be *one* solution for adding meta data to a tiddler. We currently have *tags* and *fields* and #1324 concludes to extend the *field* funcitonality. Here comes another aspect to tagging which fits in very well with my own proposals in #1324 about **unifying tags and fields** into fields, i.e `name:value` pairs but with the addition to be able to set them as visible or hidden from view mode.

Thus:
One basic way to make a *web page* turn up in internet search engines is to use "meta-tags". It is probably a *de facto* standard and has become practically self evident to do. We need something that behaves in the same way but for finding *tiddlers* in TW search. We need a way to *easily* throw in a few search tags as you're creating the tiddler, but that remain hidden from view mode. Like current tags, they should have e.g drop down list on edit for easy re-use, and we should be able to filter on them effectively. Using this feature shouldn't require more thinking than to make up the search tags. Fields are less suitable for this.

Real (and repeated) life scenario: I want to look up how to use *tags* so I go to tw.com and type type in "tags" in the search box. Search result:

>*Title matches:*
FilterOperator: tags
SystemTags
>
>*All matches:*
[long list with what appears to be less relevant search results]

As it turns out, the "correct" search term is not *tags* but *tagging*. This is anything but obvious.

Again, I refer to my descriptions in #1324, particularly towards the end.

**Comment by pmario** [comment]
TL;DR

If we need hidden content you propose, it is for TiddlyWiki only. Search engines use different standards. see comments below

--------
--------

>Thus:
One basic way to make a web page turn up in internet search engines is to use "meta-tags". It is probably a de facto standard and has become practically self evident to do. 

nope: While meta tags are still used to create additional text for the search result, they don't influence the page ranking anymore (since 2009 don't find the link atm)

see: https://support.google.com/webmasters/answer/35769?hl=en&ref_topic=6002025

The most important thing is "Content Quality"
see: second sentence in the first paragraph: 

>Even if you choose not to implement any of these suggestions, we strongly encourage you to pay very close attention to the "Quality Guidelines," which outline some of the illicit practices that may lead to a site being removed entirely from the Google index or otherwise impacted by an algorithmic or manual spam action.

---------

>We need something that behaves in the same way but for finding tiddlers in TW search. We need a way to easily throw in a few search tags as you're creating the tiddler, but that remain hidden from view mode.

no.
From "Avoid the following techniques

>Hidden Text or Links
>
>>Hiding text or links in your content to manipulate Google’s search rankings can be seen as deceptive and is a violation of Google’s Webmaster Guidelines.

-----

From "Avoid the following techniques:"

> Scraped content
>
>> Some webmasters use content taken (“scraped”) from other, more reputable sites on the assumption that increasing the volume of pages on their site is a good long-term strategy regardless of the relevance or uniqueness of that content. Purely scraped content, even from high-quality sources, may not provide any added value to your users without additional useful services or content provided by your site; it may also constitute copyright infringement in some cases. It's worthwhile to take the time to create original content that sets your site apart. This will keep your visitors coming back and will provide more useful results for users searching on Google.



---------

If we really need to to create better text for the search result page we'll need to have a look here: "Use structured data for rich search results"  https://support.google.com/webmasters/topic/4598337?hl=en&ref_topic=3309300

**Comment by twMat** [comment]
Thank you Mario! - but I merely expressed myself poorly: I really did only mean for internal TW searching and used the other things as an analogy. You info was interesting none the less, particularly to hear that meta-tags aren't used at all. I guess it was too abused to get rankings, which of course is not a concern for us.

In a way TW is like a little local internet with many webpages and when we eventually go Federation I think we will have even more use for these hidden tags. And when we one day find some system to be able to tag other peoples tiddlers, the need no be able to hide them will be even greater. But, as indicated, they will be valuable already today. The example I gave with searching for "tags" on tw.com is a real one but I particularly find in my own TWs that the more complex and filled with content of various topics, the greater the need for this.

**Comment by tobibeer** [comment]
Even for internal things, I don't see why we need extra hidden keywords.

You have 3 places to put the relevant words....

1. the title — yes, make it sound self-evident
2. the tags — after all, that's what they're for
3. the body — in a good sentence

Not enough?

**Comment by twMat** [comment]
>Even for internal things, I don't see why we need extra hidden keywords.

You say "*even* for internal things" but that is probably when the need is the *least*, and yet I say there is still a need. For *public things, the need would probably be even greater.

>the title — yes, make it sound self-evident
[...]
the body — in a good sentence
Not enough?

I obviously don't think so. It is not reasonable to demand that a user compromises the title or body for this. An obvious example is if they must follow a specific format. Besides, you could pretty much suggest "title/body" instead of tags in general. There is also the search issue: tags, be they visible or hidden, are prioritized differently than title or body to make search more effective. 

>the tags — after all, that's what they're for
As explained, some things are just not suitable for view mode. Should be obvious for public TWs but I find it even among my private ones. I could give an *unlimited* amount of examples but anyone could disagree on each single one if he really wanted to since he might simply *prefer* differently. Still, I'll try - here come Mr. Carroll again ;-)

title: Alice
visible tags: *<list of book titles>, protagonist, fictional character*
hidden tags: *Barnes&Noble deal", "check this tid if I write sequel Alice *advances* in Wonderland", "on daughters wish list", "and here is some ugly tag that will distract your tiddler reading experience", "a permanent tag so I can use this particularly suitable tiddler in my filter tests"*
text: *"some text about Alice where the hidden tags just don't fit for, say, aesthetic reasons"*

In my fictional case here, you'll just have to trust me that I don't want any of those tags visible - and this is even for *several* different reasons. And note that they are not unrealistic tags.

What would you do here? ... Aha, ok I hear you ;-) and are you saying that wouldn't compromise the tiddler itself or your workflow?

**Comment by pmario** [comment]
>You info was interesting none the less, particularly to hear that meta-tags aren't used at all. I guess it was too abused to get rankings, which of course is not a concern for us.

In an ideal world the `<meta> tag` would have been useful. But as you wrote, it has been abused to trick search engines, so they still use it but not for page ranking. eg: `<meta name="description" content="some text to be shown" />` can be used to add useful text to the search result. .... 

Adding such meta information is possible if you create a new template for a static site. ... Reading you post I actually thought, you'd like to improve TW "findability" for search engines. ... So imo for static TW created pages, it still makes sense.

**Comment by pmario** [comment]
There is a plugin (that I don't find atm), that adds "tagging" behaviour and UI for fields. Fields are "kind of" hidden by the standard view / edit templates. 

Adding "hidden tags" imo is a "no option" but  giving fields the possibility to be used more like tags imo is.

**Comment by tobibeer** [comment]
> It is not reasonable to demand that a user compromises the title or body for this.

This is de-facto how search engines work.
No meaningful content — no search results.

> There is also the search issue: tags, be they visible or hidden, are prioritized differently than title or body to make search more effective.

What "issue" are you referring to?

> Some things are just not suitable for view mode.

If they're not suitable for viewing, then they're not suitable for searching either, imho.

> I could give an unlimited amount of examples but anyone could disagree on each single one if he really wanted to since he might simply prefer differently.

I'm afraid, you really have to be specific on this one.
Without those examples, the need for this is pretty unclear.

> hidden tags: "Barnes&Noble deal", "check this tid if I write sequel Alice Advances in Wonderland", "on daughters wish list", "and here is some ugly tag that will distract your tiddler reading experience", "a permanent tag so I can use this particularly suitable tiddler in my filter tests"*

simply use...

```
<!-- my hiden keywords -->
```

Are you perhaps mixing two things? One is keeping notes for you and your eyes to see only ...so there you want to see those notes! ...and the other is publishing things for others to consume.

So, the actual requirement seems to be...

### "let me see stuff but prevent others from seeing (or finding!) it"

> In my fictional case here, you'll just have to trust me that I don't want any of those tags visible - and this is even for several different reasons. And note that they are not unrealistic tags.

But then, you find results but don't see why you found them.
Doesn't sound like a practice that I wouldn't recommend to anyone.

**Comment by twMat** [comment]
Hehe, first I just want to say thanks for always adding other perspectives guys.

@pmario 
Regarding actually adding findability in internet search engines - that *is* a very interesting (other) idea, and is currently [discussed](https://groups.google.com/forum/?fromgroups=#!topic/tiddlywikidev/6mu8wuVfzqM).

@tobibeer 
>"let me see stuff but prevent others from seeing (or finding!) it"

I want to be able to add meta-data that is (a)not visible in viewmode (b) easily added and (c) easily found in search by me but, true, not by others.

But can I ask; why *do* you want the (current) *fields* to stay hidden? I think you see what I'm getting at; sometimes you do want to hide things from view mode.

**Comment by tobibeer** [comment]
> But can I ask; why do you want the (current) fields to stay hidden? I think you see what I'm getting at; sometimes you do want to hide things from view mode.

Because fields are more programatic things that you can easily expose in view-mode via conditional templates. We don't want to see all fields all the time... because then we would have to declare what we don't want to see.

**Comment by twMat** [comment]
Yes, (and I know of course) and so regarding tags I say:

[...] that you ~~can~~ should to be able to easily ~~expose~~ hide in view-mode [ ...] We don't want to see all ~~fields~~ tags all the time.

I think we can conclude that we simply have different needs in this issue.  I'll bring it up on the boards and see how the others feel. Who knows, maybe I'm the only one.

**Comment by tobibeer** [comment]
I don't understand at all why you call them tags. These are not tags at all in a TiddlyWiki sense...

>hidden tags: *"Barnes&Noble deal", "check this tid if I write sequel Alice Advances in Wonderland", "on daughters wish list", "and here is some ugly tag that will distract your tiddler reading experience", "a permanent tag so I can use this particularly suitable tiddler in my filter tests"*

At best, it's hidden content.

You could have an additional text-editor for a second "text" field, e.g. a "hidden-text" that is only shown when some checkbox is ticked in the control panel.

**Comment by tobibeer** [comment]
Instead, I would see a benefit in this:

There is a global filter one can define which constrains *all* other filters (perhaps not in system tiddlers).

That way you could easily define something like "realms" and when switching realms you only see tiddlers pertaining to a realm... being undefined would mean the default "all".

**Comment by twMat** [comment]
>These are not tags at all in a TiddlyWiki sense.

So what are tags 'in a TiddlyWiki sense'?

And what would you say to someone who says "Well... they *are* tags because I use them as such."? You and I can't think of the applications and contexts people use/will use TW for. 

While I in real life use more compact tags, partly *because" of conforming to TW limitations, I made the example here verbose to explain them for readers of this thread. Are they better like this:

hidden tags: "B&N deal", "AAiW project", "daughters wishlist", "foobar"
...and [here](http://www.goodreads.com/list/show/2667.Books_with_Really_Long_Titles) are very conceivable tiddlertitles that could therefore be tags, and that would be very distracting (to me) if visible.

But look at the initial real life example I gave; I was searching for "tags" on tw.com. To me, "tags" is a given tag for that tiddler (the tiddler Tagging). But I wouldn't want it to be seen because it is banal and it would steal attention from the tags that I *do* want viewers to see. What do you suggest in this case? (Really what, I'm curous.)

*Of course* there are other "ways" to hide whatever-you-want-to-call-this-examples, but a key aspect with tags is IMO that they're quick and easy to add.

**Comment by tobibeer** [comment]
>  What do you suggest in this case? 

I think there are different sides to this issue.
One side is to not show certain tags as tag buttons, as in...

http://untagged.tiddlyspace.com

In TWc there were **excludeLists**, **excludeSearch**, **excludeMissing** to signal that a tiddler was not to be listed in a given spot, perhaps lacking an additional **excludeTags**.

However, in the OP you're throwing in meta-tags, websearch, fields as tags, all kinds of topics that make it difficult to grasp what this issue truly is about.

Is it about hiding distractions at a precise spot, like a tiddler's tag list? Is it about hiding them from other places, too? What does this have to do with, say, Google?

The way the OP is worded does not quite make it actionable. So, what precisely are you after? Perhaps have the title reflect that more specifically.

So, forgive me if "I have this very long sentence I wanted to throw in" does not resonate under "search tag" with me. I can see how someone might end up tagging one tiddler with another title of that length, for lack of a different, perhaps better method, e.g. a link in the tiddler body.

**Comment by twMat** [comment]
In deed, I (now) understand my OP is very fuzzy.

A start from scratch again:
* This only concerns pills/buttons in the tiddler view mode "tags area", i.e not lists or other places:
* In this tags area of a tiddler, control which tags are to be shown and which to not show
* the tags to not show are, just like the visible tags, created at the spur of the moment, i.e they are not predefined
* a tag hidden from view in tiddler A can simultaneously be visible in tiddler B

The previous reference to the Tagging tiddler on tw.com illustrates the problem. For (in this case) searchability, I would prefer it was also tagged *tags* (or perhaps *tag* if that also "covers" for *tags* in search) but this tag should not be seen when looking at the tiddler because it's a banal and would be distracting - for that particular tiddler. The tag *tags* should not be globally hidden. In all other respects it should behave like a normal tag.

I must (embarassingly) admit that I had not even considered tag lists or other contexts in my OP. For *this* proposition here, tags in lists are not concerned and should behave like currently. (It is, however, obvious to me that some kind of excludeLists command will be needed the more TW is used for public sites. And I wouldn't mind some excludeLists feature for some of my own tags either (such as one-timer tags) but that's another matter/issue.)

My fuzziness is likely because I'm not a programmer so my language is not built with the same building blocks. There are probably both benefits and drawbacks to this. I prefer to think it allows me to think out of the box more easily (...even if there are definitely some guys who can both program *and* do very original thinking here ;-)

**Comment by pmario** [comment]
> Of course there are other "ways" to hide whatever-you-want-to-call-this-examples, but a key aspect with tags is IMO that they're quick and easy to add.

I think this is the main problem here. ... Tags have a very effective / usable and most important **built in** UI. All the other field based possibilities to show / hide content, need new templates and may be a lot of work for the author.

Because of the easy to use UI authors want to use tags for things they are not built for. eg: hide them from end users, because they are (kind of) destructing. ... That's why **system tiddlers** have been invented. 

 - System tiddlers have a special prefix: `$:/`. 
 - System tiddlers are not visible in standard lists. (All, Recent, Tags, search results)
 - System tiddlers can be used as tags eg: `$:/tags/Stylesheet` 
 - .
 - If a tiddler is tagged `$:/tags/Stylesheet` it has a special meaning.
 - If a tiddler is tagged with a system tag, the system tag pill is visible, since the core devs say: It has a special meaning, that influences the wiki behaviour, that's why it must be visible.

**Content Authors say:** "These system tags are ugly and destructing / confusing for my end users." ... 

And here comes the desire to completely hide elements, that are **essential** for the wiki to work, but should be invisible for end users. ... not for the author, since they need them to develop the wiki.

With TWc we had "excludeLists, excludeSearch, excludeMissing" ... now we want "excludeTags, excludeTopBar, excludeLeftSidebar, excludeRightSidebar, excludeTiddlerMenue, excludeYouNameIt, ..." because for my usecase I don't want to see them as tag pills, but I want to see them in the right sidebar. .... IMO this isn't the solution. It just makes everything complicated and the core more bloated.

## Proposed solution:

For me the only solution is a "readers theme" that doesn't show `$:/` prefixed tiddlers. ... Earlier versions of TW had a top bar, with an edit button. (Jeremy mentioned it in the HO 76). Implementing a theme, that completely removes destructing content is possible and imo useful for the whole community. 

-----

Having a **readers theme** would make it possible for you as an author to use:

title: Alice
visible tags: protagonist, fictional character
hidden tags: $:/Barnes&Noble, $:/deal, $:/check-this-tid-if-I-write-sequel-Alice-advances-in-Wonderland, $:/whishlist/daughter, $:/whishlist/son, $:/and/here/is/some/ugly/tag-that-will+distract--your--tiddler----reading-------experience", $:/tags/test/Alice

You can use this: 
 - `<<list-links [tag[$:/whishlist/daughter]]>>`
 - `<<list-links [tag[$:/tags/test/Alice]]>>`

You can tag every system tag with eg. `$:/tags/mySystemTags` so 
 - `<<list-links [tag[$:/tags/mySystemTags]]>>` gives you a nice list of all your used system tags and a reference back to the tiddlers you tagged with them ....  

done! ... no new special edge cases ... no new "excludeFromThereBecause-I-NeedItFlags" ... 

Just plain old default behaviour with a new theme, that is useful for everyone. 

just my 2 cents.

**Comment by aelocson** [comment]
@pmario makes some good points here.

Note also that it's extremely easy to hide certain tags from the tag-pill area of a tiddler, just by changing the filter used in the view template, e.g. to `[all[current]tags[]!is[system]sort[title]]` or `[all[current]tags[]!field:pill[no]sort[title]]`.

This doesn't necessarily require a dedicated theme. We could just have a new option in the Control Panel to specify a filter identifying which tags should be shown as pills, perhaps with a shortcut button that would fill in the frequently requested case of excluding system tags.

**Comment by tobibeer** [comment]
At each tiddler that you don't want to see as a tag, rather than...

* specifying some field 
* setting some special tag

...you could simply list all tags that you want to hide in the text of a tiddler called **$:/config/hide/tag** and then use the following in **$:/core/ui/ViewTemplate/tags**...

```
<div class="tc-tags-wrapper">
<$list filter="[all[current]tags[]]-[list[$:/config/hide/tags-button!!text]]+[sort[title]]"
template="$:/core/ui/TagTemplate"
storyview="pop"/>
</div>
```

**Comment by pmario** [comment]
>Note also that it's extremely easy to hide certain tags from the tag-pill area of a tiddler,

That's right, but generally hiding information imo is a problem. If you are in author mode and still hide the tag: `$:/tags/Stylesheet` from the tag list. ... Someone may post to the google group, that they have a tiddler that is named: myFreekingStyles. .. It contains CSS info and is not tagged with a system tiddler. ... Why does it work? .... That's exactly the problem the "excludeXXXX" flag caused with TWc. ....

It causes support requests and wastes time, tracking down, very hard to find issues.

**Comment by tobibeer** [comment]
>  It contains CSS info and is not tagged with a system tiddler

So long as it's visible in edit-mode, I don't see a problem.

**Comment by twMat** [comment]
Thank you Mario. I see your point and the question with core vs theme/plugin/etc highlights the difficulty to know what the needs actually are, i.e how many that would benefit from something so to know if it is justified to consider for core. (...that *is* factored in, right?)

However, I just posted on the boards about [The Concept of Tags and Tagging](https://groups.google.com/forum/?fromgroups#!topic/tiddlywiki/iNVgY_NdVTQ) outlining a perspective and an explanation why I see reason to rethink how tagging functions. Tagging is simply a very powerful tool but we are not taking advantage of its full potential, IMO.
 
@pmario 
Regarding hiding things by using system tiddler notation; this is a great idea, which I will use (thank you!), but it does not address the problems outlined here as should be clarified in the above referenced the google groups thread. It is really about primarily hiding tags in the actual tiddler view (in view mode). 

@tobibeer - and as hopefully also is clear inthe above referenced post, predefined lists is a great solution, but to another problem.

I think @aelocson is closer to a practical solution but I'm hoping for a very simple interface where the user wouldn't have to create hidden tags in a minimally different way than visible tags. If we eventually do have field `name:value` pairs for tags, then I guess what I call a [semantid fieldname](https://groups.google.com/forum/?fromgroups#!topic/tiddlywiki/zUvLEXv9BAk) (is there an established term?) might be able to set a property - i.e what I mean with a "semantic fieldname" is a fieldname that is not only a type attribute for a value, but that the fieldname *itself* is interpreted as a command. An example might be "hidden:foo" meaning that "foo" is a of the type hidden and if this is the case then something *happens* to the value, such as "display the value, using this template". (Impossible ?)

A simpler solution, if it Is possible, might be a second tags field but that is simply not shown in view mode? I.e it is appended(?) to the first tags field so that the contents (i.e the tags) are in deed treated like tags in all aspects (appearing in lists etc, searchable like tags etc).

I appreciate your input guys :-)

**Comment by tobibeer** [comment]
> predefined lists is a great solution, but to another problem.

* the list is not "predefined", it is user defined
* how tags get listed there is indeed a question
** could happen via much the same ui as the edit tags

**Comment by twMat** [comment]
>the list is not "predefined", it is user defined
how tags get listed there is indeed a question
  but a simple button to add or remove will do

Ahaaaa! Now that is interesting! Imagine this: tag pills in *edit* mode are (surprisingly) passive - all you can do is delete a tag. What if the tagpill in edit mode was itself actually such a button! Clicking it adds/removes the title to the list and also renders the tag pill color to be "invisible" (white or background) and the text black (it is by default white in edit). 
How elegant! Or would this break something?

**Comment by tobibeer** [comment]
> How elegant! Or would this break something?

This sounds indeed like a most elegant solution. However, I think this should toggle the tag visibility for all tiddlers, not just the current one! The case where I want to hide a single tag from one tiddler only is the exception, not the rule.

**Comment by aelocson** [comment]
Yes, that's a lovely idea. Very clean and simple.

**Comment by pmario** [comment]
@Jermolene ... "discussion"

**Comment by twMat** [comment]
Closing. Overall, the OP is confusing and fuzzy- **but** TW ought to feature a way to hide tags from viewmode and the idea to use the editview tag itself to set it as visibile/hidden is very good so I'll repost it as a more defined request.
